cpp/ql/src/semmle/code/cpp/models/implementations/Pure.qll

```diff
@@ -47,11 +48,60 @@ class PureFunction extends ArrayFunction, TaintFunction {
       )
     ) and
     (
+      output.isOutReturnValue() or
+      output.isOutReturnPointer()
+    )
+  }
+
+  override predicate parameterNeverEscapes(int i) {
+    getParameter(i).getUnspecifiedType() instanceof PointerType and
+    not (
+      i = 0 and
+      getType().getUnspecifiedType() instanceof PointerType
+    )
+  }
+
+  override predicate parameterEscapesOnlyViaReturn(int i) {
+    i = 0 and
+    getType().getUnspecifiedType() instanceof PointerType
+  }
+
+  override predicate parameterIsAlwaysReturned(int i) {
+    none()
+  }
+
+  override predicate neverReadsMemory() {
+    none()
+  }
+  
+  override predicate neverWritesMemory() {
+    any()
+  }
+}
+
+class PureFunction extends TaintFunction, SideEffectFunction {
+  PureFunction() {
+    exists(string name |
+      hasName(name) and
       (
-        output.isOutReturnPointer() and
-        getUnspecifiedType() instanceof PointerType
-      ) or
-      output.isOutReturnValue()
+        name = "abs" or
+        name = "labs"
+      )
     )
   }
+
+  override predicate hasTaintFlow(FunctionInput input, FunctionOutput output) {
+    exists (ParameterIndex i |
+      input.isInParameter(i)
+    ) and
```

Comment:
If there's a function in the snapshot with 20 parameters, then this predicate claims that parameter 19 of `abs` flows to its return value. That seems wasteful and confusing.

I'd much prefer that we only define taint flow for parameters that actually exist. The same issue is in `PureStrFunction`.

To make this easier to detect in the future, we could make `ParameterIndex` private and put a `bindingset` annotation on `isInParameter` etc. But that's probably not for this PR.

Reply:
I've narrowed it to only report parameters that actually exist